src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs

```diff
@@ -901,12 +997,173 @@ class A
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
+        }
+
+        [Fact]
+        public void AsyncVoidMain_CSharp71()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    static async void Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+        System.Console.WriteLine(""Async Void Main"");
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1)).VerifyDiagnostics(
+                // (6,23): warning CS0028: 'A.Main(string[])' has the wrong signature to be an entry point
+                //     static async void Main(string[] args)
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("A.Main(string[])").WithLocation(6, 23),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
+        }
+
+        [Fact]
+        public void TaskMainAndNonTaskMain_WithExplicitMain()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    static void Main()
+    {
+        System.Console.WriteLine(""Non Task Main"");
+    }
+    async static Task Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+        System.Console.WriteLine(""Task Main"");
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe.WithMainTypeName("A"), parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics();
+            CompileAndVerify(compilation, expectedOutput: "Non Task Main", expectedReturnCode: 0);
+        }
+
+        [Fact]
+        public void TaskOfFloatMainAndNonTaskMain()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    static void Main()
+    {
+        System.Console.WriteLine(""Non Task Main"");
+    }
+
+    async static Task<float> Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+        System.Console.WriteLine(""Task Main"");
+        return 0.0f;
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics(
+                // (10,30): warning CS0028: 'A.Main(string[])' has the wrong signature to be an entry point
+                //     async static Task<float> Main(string[] args)
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("A.Main(string[])").WithLocation(11, 30));
+            CompileAndVerify(compilation, expectedOutput: "Non Task Main", expectedReturnCode: 0);
```

Comment:
Yeah, I agree, but my current implementation is analogous to the previous non-task-based handling of mains.

```c#
static int main(){}
// WARNING
static float main(string[] args) {}
```

```c#
static int main(){}
// WARNING
static Task<float> main(string[] args) {}
```

```c#
static Task<int> main(){}
// WARNING
static Task<float> main(string[] args) {}
```